### PR TITLE
add file handling and uploading from services

### PR DESCRIPTION
**Pull request by Chyslovsky**

### CODE REVIEW

File: Exceptions/FileUploadException.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * @package Garlic\Bus\Exceptions
+ */
+class FileUploadException extends \Exception
+{
+    /**
+     * @var string
+     */
+    private $fileName;
+
+
```
Comment (am0nshi) [inline Exceptions/FileUploadException.php]: double

File: Exceptions/FileUploadException.php [hunk 2]
```diff
@@ -0,0 +1,33 @@
+<?php
+
+
+namespace Garlic\Bus\Exceptions;
+
+
+use Throwable;
```
Comment (am0nshi) [inline Exceptions/FileUploadException.php]: \Throwable

**Review by am0nshi — DISMISSED** [review]

File: Service/File/ScpFileUploadService.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        ssh2_scp_recv($connection, $metadata['path'], $filePath);
+        if (!file_exists($filePath)) {
+            throw new FileUploadException($metadata['path']);
+        }
+
+        return $filePath;
+    }
+
+    /**
+     * @param $metadata
```
Comment (imaximius) [inline Service/File/ScpFileUploadService.php:67]: Could you add description to this method?

**Review by imaximius — CHANGES_REQUESTED** [review]

File: Exceptions/FileUploadException.php (hunk 2 above)
Comment (Chyslovsky) [reply Exceptions/FileUploadException.php]: done

File: Exceptions/FileUploadException.php (hunk 1 above)
Comment (Chyslovsky) [reply Exceptions/FileUploadException.php]: done

File: Service/File/ScpFileUploadService.php (hunk 3 above)
Comment (Chyslovsky) [reply Service/File/ScpFileUploadService.php:67]: done

File: Service/File/FileHandlerService.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * Class FileHandlerService
+ *
+ * @package Garlic\Bus\Service\File
+ */
+class FileHandlerService
+{
+    /**
+     * @var
```
Comment (demotivationme) [inline Service/File/FileHandlerService.php]: string

File: Service/File/FileHandlerService.php [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    private $hostUrl;
+    /**
+     * @var
```
Comment (demotivationme) [inline Service/File/FileHandlerService.php]: string

File: Service/File/FileHandlerService.php (hunk 4 above)
Comment (Chyslovsky) [reply Service/File/FileHandlerService.php]: done

File: Service/File/FileHandlerService.php (hunk 5 above)
Comment (Chyslovsky) [reply Service/File/FileHandlerService.php]: done

**Review by demotivationme — DISMISSED** [review]

File: Service/File/ScpFileUploadService.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+     * @param array $metadata
+     *
+     * @return string
+     * @throws FileUploadException
+     */
+    public function getFile(array $metadata)
+    {
+        $connection = $this->getSshConnection($metadata);
+        $filePath = $this->uploadDir . md5(time() . $metadata['origin_name']);
+        ssh2_scp_recv($connection, $metadata['path'], $filePath);
```
Comment (am0nshi) [inline Service/File/ScpFileUploadService.php]: potential non-FileUploadException exception. wrap into try/catch for proper handling?

**Review by am0nshi — DISMISSED** [review]

File: Service/File/ScpFileUploadService.php (hunk 6 above)
Comment (Chyslovsky) [reply Service/File/ScpFileUploadService.php]: done

**Review by imaximius — APPROVED** [review]

**Review by demotivationme — APPROVED** [review]

**Pull request merged**